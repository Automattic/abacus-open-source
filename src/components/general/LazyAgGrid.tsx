import { LinearProgress } from '@material-ui/core'
import { AgGridReactProps } from 'ag-grid-react'
import React, { Suspense } from 'react'

const AgGrid = React.lazy(() => import('./AgGridReact'))

export default function LazyAgGrid(props: AgGridReactProps): JSX.Element {
  return (
    <Suspense fallback={<LinearProgress />}>
      <AgGrid {...props} />
    </Suspense>
  )
}
