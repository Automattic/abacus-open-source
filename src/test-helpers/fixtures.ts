/**
 * Data fixtures to use in tests.
 *
 * Functions in this file return new objects populated with dummy values, which may potentially be overridden in
 * functions that accept Partial<T> as an argument.
 */
import _ from 'lodash'

import {
  Analysis,
  AnalysisStrategy,
  AssignmentCacheStatus,
  AttributionWindowSeconds,
  DistributionStats,
  ExperimentFull,
  ExperimentFullNew,
  Metric,
  MetricAssignment,
  MetricEstimates,
  MetricNew,
  MetricParameterType,
  Platform,
  Segment,
  SegmentAssignment,
  SegmentType,
  Status,
  TagBare,
  TransactionTypes,
  Variation,
} from 'src/lib/explat/schemas'

function createDistributionStats(fieldOverrides?: Partial<DistributionStats>): DistributionStats {
  return {
    mean: 0,
    top_95: 0,
    bottom_95: 0,
    top_99: 0,
    bottom_99: 0,
    top_50: 0,
    bottom_50: 0,
    ...fieldOverrides,
  }
}

function createMetricEstimates(fieldOverrides?: Partial<MetricEstimates>): MetricEstimates {
  return {
    variations: {
      '1': createDistributionStats(),
      '2': createDistributionStats(),
    },
    diffs: {
      '1_2': createDistributionStats(),
      '2_1': createDistributionStats(),
    },
    ratios: {
      '1_2': createDistributionStats(),
      '2_1': createDistributionStats(),
    },
    ...fieldOverrides,
  }
}

function createAnalysis(fieldOverrides: Partial<Analysis>): Analysis {
  return {
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.IttPure,
    participantStats: {
      total: 1000,
      variation_1: 600,
      variation_2: 400,
    },
    metricEstimates: createMetricEstimates(),
    analysisDatetime: new Date(Date.UTC(2020, 4, 10)),
    ...fieldOverrides,
  }
}

function createMultiVariationAnalysis(fieldOverrides: Partial<Analysis>): Analysis {
  return {
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.PpNaive,
    participantStats: {
      total: 1000,
      variation_1: 400,
      variation_2: 400,
      variation_3: 200,
    },
    metricEstimates: {
      variations: {
        '1': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: 0.5,
          mean: 1,
        }),
        '2': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: 0.5,
          mean: 1,
        }),
        '3': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: 0.5,
          mean: 1,
        }),
      },
      diffs: {
        '2_1': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: -1,
          mean: 0,
        }),
        '1_2': Fixtures.createDistributionStats({
          top_95: 0,
          bottom_95: 0,
          mean: 0,
        }),
        '3_1': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: -1,
          mean: 0,
        }),
        '1_3': Fixtures.createDistributionStats({
          top_95: 0,
          bottom_95: 0,
          mean: 0,
        }),
        '3_2': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: -1,
          mean: 0,
        }),
        '2_3': Fixtures.createDistributionStats({
          top_95: 0,
          bottom_95: 0,
          mean: 0,
        }),
      },
      ratios: {
        '2_1': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: 0.5,
          mean: 0,
        }),
        '1_2': Fixtures.createDistributionStats({
          top_95: 0,
          bottom_95: 0,
          mean: 0,
        }),
        '3_1': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: 0.5,
          mean: 0,
        }),
        '1_3': Fixtures.createDistributionStats({
          top_95: 0,
          bottom_95: 0,
          mean: 0,
        }),
        '3_2': Fixtures.createDistributionStats({
          top_95: 1,
          bottom_95: 0.5,
          mean: 0,
        }),
        '2_3': Fixtures.createDistributionStats({
          top_95: 0,
          bottom_95: 0,
          mean: 0,
        }),
      },
    },
    analysisDatetime: new Date(Date.UTC(2020, 4, 10)),
    ...fieldOverrides,
  }
}

function createAnalyses(): Analysis[] {
  return [
    // Full set of "latest" analyses for the default metric assignment.
    createAnalysis({
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 1000,
        variation_1: 600,
        variation_2: 400,
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoCrossovers,
      participantStats: {
        total: 900,
        variation_1: 540,
        variation_2: 360,
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammers,
      participantStats: {
        total: 850,
        variation_1: 510,
        variation_2: 340,
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
      participantStats: {
        total: 800,
        variation_1: 480,
        variation_2: 320,
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 700,
        variation_1: 420,
        variation_2: 280,
      },
    }),

    // Another set of analyses for the default metric assignment with an earlier date.
    createAnalysis({
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 100,
        variation_1: 60,
        variation_2: 40,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoCrossovers,
      participantStats: {
        total: 90,
        variation_1: 54,
        variation_2: 36,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammers,
      participantStats: {
        total: 85,
        variation_1: 51,
        variation_2: 34,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
      participantStats: {
        total: 80,
        variation_1: 48,
        variation_2: 32,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 70,
        variation_1: 42,
        variation_2: 28,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),

    // One example of a metric assignment with no data on one variation.
    createAnalysis({
      metricAssignmentId: 124,
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 10,
        variation_1: 10,
      },
      metricEstimates: null,
    }),

    // Similar to the set of "latest" analyses for the default metric assignment, but with consistent recommendations.
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 2000,
        variation_1: 1200,
        variation_2: 800,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.MittNoCrossovers,
      participantStats: {
        total: 1800,
        variation_1: 1080,
        variation_2: 720,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.MittNoSpammers,
      participantStats: {
        total: 1700,
        variation_1: 920,
        variation_2: 780,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
      participantStats: {
        total: 1600,
        variation_1: 960,
        variation_2: 640,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 1400,
        variation_1: 840,
        variation_2: 560,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 1200,
        variation_1: 940,
        variation_2: 660,
      },
      analysisDatetime: new Date(Date.UTC(2020, 5, 11)),
    }),
  ]
}

function createMetricAssignment(fieldOverrides: Partial<MetricAssignment>): MetricAssignment {
  return {
    metricAssignmentId: 123,
    metricId: 1,
    attributionWindowSeconds: AttributionWindowSeconds.OneWeek,
    changeExpected: true,
    isPrimary: true,
    minDifference: 0.1,
    ...fieldOverrides,
  }
}

/* istanbul ignore next; All coverage to be removed, see https://github.com/Automattic/abacus/issues/231 */
function createExperimentFullNew(fieldOverrides: Partial<ExperimentFullNew> = {}): ExperimentFullNew {
  const now = new Date()
  // This is a bit funky as we want to work in UTC but DateFns only does local time
  // and we want to avoid DST problems.
  const startDatetime = new Date(new Date(new Date().setMonth(now.getMonth() + 2)).setUTCHours(0, 0, 0, 0))
  const endDatetime = new Date(new Date(new Date().setMonth(now.getMonth() + 4)).setUTCHours(0, 0, 0, 0))
  return {
    name: 'experiment_1',
    startDatetime,
    endDatetime,
    platform: Platform.Calypso,
    ownerLogin: 'owner-nickname',
    description: 'Experiment with things. Change stuff. Profit.',
    existingUsersAllowed: false,
    p2Url: 'https://wordpress.com/experiment_1',
    variations: [
      {
        name: 'test',
        isDefault: false,
        allocatedPercentage: 40,
      },
      {
        name: 'control',
        isDefault: true,
        allocatedPercentage: 60,
      },
    ],
    metricAssignments: [
      createMetricAssignment({
        metricAssignmentId: 123,
        metricId: 1,
        attributionWindowSeconds: AttributionWindowSeconds.OneWeek,
        changeExpected: true,
        isPrimary: true,
        minDifference: 0.1,
      }),
      createMetricAssignment({
        metricAssignmentId: 124,
        metricId: 2,
        attributionWindowSeconds: AttributionWindowSeconds.FourWeeks,
        changeExpected: false,
        isPrimary: false,
        minDifference: 10.5,
      }),
      createMetricAssignment({
        metricAssignmentId: 125,
        metricId: 2,
        attributionWindowSeconds: AttributionWindowSeconds.OneHour,
        changeExpected: true,
        isPrimary: false,
        minDifference: 0.5,
      }),
      createMetricAssignment({
        metricAssignmentId: 126,
        metricId: 3,
        attributionWindowSeconds: AttributionWindowSeconds.SixHours,
        changeExpected: true,
        isPrimary: false,
        minDifference: 12,
      }),
    ],
    segmentAssignments: [],
    ...fieldOverrides,
  }
}

function createVariation(fieldOverrides: Partial<Variation> = {}): Variation {
  const variationId = fieldOverrides.variationId || 1
  const name = variationId === 1 ? 'control' : variationId === 2 ? 'test' : `treatment_${variationId - 1}`
  return {
    variationId,
    isDefault: variationId === 1 ? true : false,
    name,
    allocatedPercentage: 10,
    ...fieldOverrides,
  }
}

function createVariations(n: number): Variation[] {
  const allocatedPercentage = Math.floor(100 / n)
  return _.range(1, n + 1).map((variationId) =>
    createVariation({
      variationId,
      allocatedPercentage,
    }),
  )
}

function createExperimentFull(fieldOverrides: Partial<ExperimentFull> = {}): ExperimentFull {
  const fieldsOnlyForExistingExperiments = [
    'experimentId',
    'endReason',
    'conclusionUrl',
    'deployedVariationId',
    'status',
    'variations',
    'metricAssignments',
    'segmentAssignments',
    'exposureEvents',
    'assignmentCacheStatus',
  ]
  const newExperimentFieldOverrides = {
    ..._.omit(fieldOverrides, fieldsOnlyForExistingExperiments),
    exposureEvents: undefined,
    status: undefined,
    assignmentCacheStatus: undefined,
  }
  const existingExperimentFieldOverrides = _.pick(fieldOverrides, fieldsOnlyForExistingExperiments)

  return {
    ...createExperimentFullNew(newExperimentFieldOverrides),
    experimentId: 1,
    status: Status.Staging,
    endReason: null,
    conclusionUrl: null,
    deployedVariationId: null,
    variations: [
      createVariation({
        variationId: 2,
        name: 'test',
        isDefault: false,
        allocatedPercentage: 40,
      }),
      createVariation({
        allocatedPercentage: 60,
      }),
    ],
    metricAssignments: [
      createMetricAssignment({
        metricAssignmentId: 123,
        metricId: 1,
        attributionWindowSeconds: AttributionWindowSeconds.OneWeek,
        changeExpected: true,
        isPrimary: true,
        minDifference: 0.1,
      }),
      createMetricAssignment({
        metricAssignmentId: 124,
        metricId: 2,
        attributionWindowSeconds: AttributionWindowSeconds.FourWeeks,
        changeExpected: false,
        isPrimary: false,
        minDifference: 10.5,
      }),
      createMetricAssignment({
        metricAssignmentId: 125,
        metricId: 2,
        attributionWindowSeconds: AttributionWindowSeconds.OneHour,
        changeExpected: true,
        isPrimary: false,
        minDifference: 0.5,
      }),
      createMetricAssignment({
        metricAssignmentId: 126,
        metricId: 3,
        attributionWindowSeconds: AttributionWindowSeconds.SixHours,
        changeExpected: true,
        isPrimary: false,
        minDifference: 12,
      }),
    ],
    segmentAssignments: [
      {
        segmentAssignmentId: 1,
        segmentId: 1,
        isExcluded: true,
      },
    ],
    exposureEvents: [
      {
        event: 'event_name',
        props: {
          additionalProp1: 'prop1Value',
          additionalProp2: 'prop2Value',
          additionalProp3: 'prop3Value',
        },
      },
      {
        event: 'event_without_props',
      },
    ],
    exclusionGroupTagIds: [1],
    assignmentCacheStatus: AssignmentCacheStatus.Fresh,
    ...existingExperimentFieldOverrides,
  }
}

function createMetric(id: number, override?: Partial<Metric>): Metric {
  // Note: It is hard to reuse createMetric here as it is boxed
  //       Currently we only unbox it into an ApiData format which is different from this
  const parameterType = id % 2 === 0 ? MetricParameterType.Revenue : MetricParameterType.Conversion
  const eventParams = [{ event: 'event_name', props: { has_blocks: 'true' } }]
  const revenueParams = {
    refundDays: id * 2,
    productSlugs: ['xx-bundles'],
    transactionTypes: [TransactionTypes.NewPurchase],
  }
  return {
    metricId: id,
    name: `metric_${id}`,
    description: `This is metric ${id}`,
    parameterType,
    higherIsBetter: id % 3 === 0,
    eventParams: parameterType === MetricParameterType.Conversion ? eventParams : undefined,
    revenueParams: parameterType === MetricParameterType.Revenue ? revenueParams : undefined,
    tags: [],
    ...override,
  }
}

function createMetricNew(id: number, override?: Partial<Metric>): MetricNew {
  const metric = createMetric(id, override)
  return { ...metric, tags: metric.tags?.map((tag) => tag.tagId) || [] }
}

function createMetrics(numMetrics = 3): Metric[] {
  return _.range(1, numMetrics + 1).map((id) => createMetric(id))
}

function createTagBare(id: number): TagBare {
  return {
    tagId: id,
    namespace: `tag_namespace_${id}`,
    name: `tag_${id}`,
    description: `This is tag ${id}`,
  }
}

function createTagBares(numTags = 3): TagBare[] {
  return _.range(1, numTags + 1).map(createTagBare)
}

const createTagFull = createTagBare

function createSegment(id: number): Segment {
  return {
    segmentId: id,
    name: `segment_${id}`,
    type: id % 2 === 0 ? SegmentType.Country : SegmentType.Locale,
  }
}

/**
 * Creates an array of segments.
 */
function createSegments(numSegments = 2): Segment[] {
  return _.range(1, numSegments + 1).map(createSegment)
}

function createSegmentAssignment(fieldOverrides: Partial<SegmentAssignment>): SegmentAssignment {
  return {
    segmentAssignmentId: 123,
    segmentId: 1,
    isExcluded: false,
    ...fieldOverrides,
  }
}

const Fixtures = {
  createAnalyses,
  createAnalysis,
  createMultiVariationAnalysis,
  createDistributionStats,
  createExperimentFull,
  createExperimentFullNew,
  createMetricAssignment,
  createMetricEstimates,
  createMetric,
  createMetrics,
  createMetricNew,
  createTagBares,
  createTagFull,
  createSegmentAssignment,
  createSegments,
  createVariation,
  createVariations,
}

export default Fixtures
