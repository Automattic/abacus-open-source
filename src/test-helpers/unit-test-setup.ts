import * as notistack from 'notistack'

/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
// @ts-ignore
window.URL.createObjectURL = jest.fn()

window.HTMLCanvasElement.prototype.getContext = jest.fn()

jest.mock('notistack')
const mockedNotistack = notistack as jest.Mocked<typeof notistack>
beforeEach(() => {
  mockedNotistack.useSnackbar.mockImplementation(() => ({
    enqueueSnackbar: jest.fn(),
    closeSnackbar: jest.fn(),
  }))
})

global.console = {
  ...console,
  error: jest.fn(),
}
