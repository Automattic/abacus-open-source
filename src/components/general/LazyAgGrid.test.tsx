import { screen } from '@testing-library/react'
import React from 'react'

import { render } from 'src/test-helpers/test-utils'

import LazyAgGrid from './LazyAgGrid'

test('renders progress loading bar', async () => {
  render(<LazyAgGrid />)

  expect(await screen.findByRole('progressbar'))
})

test('renders a grid', async () => {
  const { container } = render(<LazyAgGrid />)

  expect(container).toMatchSnapshot()
})
