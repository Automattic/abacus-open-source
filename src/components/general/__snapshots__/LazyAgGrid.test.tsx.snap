// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a grid 1`] = `
<div>
  <div
    style="height: 100%;"
  >
    <div
      class="ag-root-wrapper ag-layout-normal ag-ltr"
      ref="eRootWrapper"
    >
      
                
                
      <div
        class="ag-root-wrapper-body ag-layout-normal ag-focus-managed"
        ref="rootWrapperBody"
      >
        <div
          class="ag-tab-guard ag-tab-guard-top"
          role="presentation"
          tabindex="0"
        />
        
                    
        <!--AG-GRID-COMP-->
        <div
          aria-colcount="0"
          aria-rowcount="1"
          class="ag-root ag-unselectable ag-layout-normal"
          ref="gridPanel"
          role="grid"
          unselectable="on"
        >
          
        
          <!--AG-HEADER-ROOT-->
          <div
            class="ag-header ag-focus-managed ag-pivot-off"
            ref="headerRoot"
            role="presentation"
            style="height: 26px; min-height: 26px;"
            unselectable="on"
          >
            
            
            <div
              class="ag-pinned-left-header ag-hidden"
              ref="ePinnedLeftHeader"
              role="presentation"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            >
              <div
                aria-rowindex="1"
                class="ag-header-row ag-header-row-column"
                role="row"
                style="top: 0px; height: 25px; width: 0px;"
              />
            </div>
            
            
            <div
              class="ag-header-viewport"
              ref="eHeaderViewport"
              role="presentation"
            >
              
                
              <div
                class="ag-header-container"
                ref="eHeaderContainer"
                role="rowgroup"
                style="width: 0px;"
              >
                <div
                  aria-rowindex="1"
                  class="ag-header-row ag-header-row-column"
                  role="row"
                  style="top: 0px; height: 25px; width: 0px;"
                />
              </div>
              
            
            </div>
            
            
            <div
              class="ag-pinned-right-header ag-hidden"
              ref="ePinnedRightHeader"
              role="presentation"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            >
              <div
                aria-rowindex="1"
                class="ag-header-row ag-header-row-column"
                role="row"
                style="top: 0px; height: 25px; width: 0px;"
              />
            </div>
            
        
          </div>
          
        
          <div
            class="ag-floating-top"
            ref="eTop"
            role="presentation"
            style="min-height: 0px; height: 0px; display: none; overflow-y: hidden;"
            unselectable="on"
          >
            
            
            <div
              class="ag-pinned-left-floating-top ag-hidden"
              ref="eLeftTop"
              role="presentation"
              unselectable="on"
            />
            
            
            <div
              class="ag-floating-top-viewport"
              ref="eTopViewport"
              role="presentation"
              unselectable="on"
            >
              
                
              <div
                class="ag-floating-top-container"
                ref="eTopContainer"
                role="presentation"
                style="width: 0px;"
                unselectable="on"
              />
              
            
            </div>
            
            
            <div
              class="ag-pinned-right-floating-top ag-hidden"
              ref="eRightTop"
              role="presentation"
              unselectable="on"
            />
            
            
            <div
              class="ag-floating-top-full-width-container ag-hidden"
              ref="eTopFullWidthContainer"
              role="presentation"
              unselectable="on"
            />
            
        
          </div>
          
        
          <div
            class="ag-body-viewport ag-layout-normal ag-row-no-animation"
            ref="eBodyViewport"
            role="presentation"
          >
            
            
            <div
              class="ag-pinned-left-cols-container ag-hidden"
              ref="eLeftContainer"
              role="presentation"
              style="height: 1px;"
              unselectable="on"
            />
            
            
            <div
              class="ag-center-cols-clipper"
              ref="eCenterColsClipper"
              role="presentation"
              style="height: 1px;"
              unselectable="on"
            >
              
                
              <div
                class="ag-center-cols-viewport"
                ref="eCenterViewport"
                role="presentation"
                style="height: calc(100% + 0px);"
              >
                
                    
                <div
                  class="ag-center-cols-container"
                  ref="eCenterContainer"
                  role="rowgroup"
                  style="width: 0px; height: 1px;"
                  unselectable="on"
                />
                
                
              </div>
              
            
            </div>
            
            
            <div
              class="ag-pinned-right-cols-container ag-hidden"
              ref="eRightContainer"
              role="presentation"
              style="height: 1px;"
              unselectable="on"
            />
            
            
            <div
              class="ag-full-width-container"
              ref="eFullWidthContainer"
              role="presentation"
              style="height: 1px;"
              unselectable="on"
            />
            
        
          </div>
          
        
          <div
            class="ag-floating-bottom"
            ref="eBottom"
            role="presentation"
            style="min-height: 0px; height: 0px; display: none; overflow-y: hidden;"
            unselectable="on"
          >
            
            
            <div
              class="ag-pinned-left-floating-bottom ag-hidden"
              ref="eLeftBottom"
              role="presentation"
              unselectable="on"
            />
            
            
            <div
              class="ag-floating-bottom-viewport"
              ref="eBottomViewport"
              role="presentation"
              unselectable="on"
            >
              
                
              <div
                class="ag-floating-bottom-container"
                ref="eBottomContainer"
                role="presentation"
                style="width: 0px;"
                unselectable="on"
              />
              
            
            </div>
            
            
            <div
              class="ag-pinned-right-floating-bottom ag-hidden"
              ref="eRightBottom"
              role="presentation"
              unselectable="on"
            />
            
            
            <div
              class="ag-floating-bottom-full-width-container ag-hidden"
              ref="eBottomFullWidthContainer"
              role="presentation"
              unselectable="on"
            />
            
        
          </div>
          
        
          <div
            aria-hidden="true"
            class="ag-body-horizontal-scroll"
            ref="eHorizontalScrollBody"
            style="height: 0px; max-height: 0px; min-height: 0px;"
          >
            
            
            <div
              class="ag-horizontal-left-spacer"
              ref="eHorizontalLeftSpacer"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            
            
            <div
              class="ag-body-horizontal-scroll-viewport"
              ref="eBodyHorizontalScrollViewport"
              style="height: 0px; max-height: 0px; min-height: 0px;"
            >
              
                
              <div
                class="ag-body-horizontal-scroll-container"
                ref="eBodyHorizontalScrollContainer"
                style="width: 0px; height: 0px; max-height: 0px; min-height: 0px;"
              />
              
            
            </div>
            
            
            <div
              class="ag-horizontal-right-spacer"
              ref="eHorizontalRightSpacer"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            
        
          </div>
          
        
          <!--AG-OVERLAY-WRAPPER-->
          <div
            aria-hidden="true"
            class="ag-overlay"
            ref="overlayWrapper"
          >
            
            
            <div
              class="ag-overlay-panel"
            >
              
                
              <div
                class="ag-overlay-wrapper ag-layout-normal ag-overlay-loading-wrapper"
                ref="eOverlayWrapper"
              >
                <span
                  class="ag-overlay-loading-center"
                >
                  Loading...
                </span>
              </div>
              
            
            </div>
            
        
          </div>
          
    
        </div>
        
                    
                
        <div
          class="ag-tab-guard ag-tab-guard-bottom"
          role="presentation"
          tabindex="0"
        />
      </div>
      
                
                
      <!--AG-PAGINATION-->
      <div
        aria-describedby="ag-5-start-page ag-5-start-page-number ag-5-of-page ag-5-of-page-number ag-5-first-row ag-5-to ag-5-last-row ag-5-of ag-5-row-count"
        aria-live="polite"
        class="ag-paging-panel ag-unselectable ag-hidden"
        id="ag-5"
      >
        
                
        <span
          aria-hidden="true"
          class="ag-paging-row-summary-panel"
        >
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            id="ag-5-first-row"
            ref="lbFirstRowOnPage"
          />
          
                    
          <span
            id="ag-5-to"
          >
            to
          </span>
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            id="ag-5-last-row"
            ref="lbLastRowOnPage"
          />
          
                    
          <span
            id="ag-5-of"
          >
            of
          </span>
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            id="ag-5-row-count"
            ref="lbRecordCount"
          />
          
                
        </span>
        
                
        <span
          class="ag-paging-page-summary-panel"
          role="presentation"
        >
          
                    
          <div
            aria-label="First Page"
            class="ag-paging-button"
            ref="btFirst"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-first"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <div
            aria-label="Previous Page"
            class="ag-paging-button"
            ref="btPrevious"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-previous"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <span
            aria-hidden="true"
            class="ag-paging-description"
          >
            
                        
            <span
              id="ag-5-start-page"
            >
              Page
            </span>
            
                        
            <span
              class="ag-paging-number"
              id="ag-5-start-page-number"
              ref="lbCurrent"
            />
            
                        
            <span
              id="ag-5-of-page"
            >
              of
            </span>
            
                        
            <span
              class="ag-paging-number"
              id="ag-5-of-page-number"
              ref="lbTotal"
            />
            
                    
          </span>
          
                    
          <div
            aria-label="Next Page"
            class="ag-paging-button"
            ref="btNext"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-next"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <div
            aria-label="Last Page"
            class="ag-paging-button"
            ref="btLast"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-last"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                
        </span>
        
            
      </div>
      
                
            
    </div>
  </div>
</div>
`;
